OpenRA.Game/Traits/World/MusicPlaylist.cs

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.GameRules;
+
+namespace OpenRA.Traits
+{
+	[Desc("Trait for music handling. Attach this to the world actor.")]
+	public class MusicPlaylistInfo : ITraitInfo
+	{
+		public readonly string StartingMusic = null;
+		public readonly bool LoopStartingMusic = false;
+
+		public object Create(ActorInitializer init) { return new MusicPlaylist(init.World, this); }
+	}
+
+	public class MusicPlaylist
+	{
+		bool installed = false;
+		bool repeat;
+
+		MusicInfo currentSong = null;
+		MusicInfo[] music;
+		MusicInfo[] random;
+
+		public MusicInfo[] GetPlaylist { get { return music; } }
+		public bool GetInstalled { get { return installed; } }
+
+		public MusicPlaylist(World world, MusicPlaylistInfo info)
+		{
+			installed = world.Map.Rules.InstalledMusic.Any();
+
+			music = world.Map.Rules.InstalledMusic.Select(a => a.Value).ToArray();
+			if (!installed)
+				return;
+
+			random = music.Shuffle(Game.CosmeticRandom).ToArray();
+
+			if (Game.Settings.Sound.MapMusic && !string.IsNullOrEmpty(info.StartingMusic) && world.Map.Rules.Music.ContainsKey(info.StartingMusic))
+			{
+				currentSong = world.Map.Rules.Music[info.StartingMusic];
+				repeat = info.LoopStartingMusic;
+			}
+			else
+			{
+				currentSong = Game.Settings.Sound.Shuffle ? random.First() : music.First();
+				repeat = Game.Settings.Sound.Repeat;
+			}
+
+			Play();
+		}
+
+		void Play()
+		{
+			if (currentSong == null)
+				return;
+
+			Sound.PlayMusicThen(currentSong, () =>
+			{
+				if (!installed)
+					return;
+
+				if (!repeat)
+					currentSong = GetNextSong();
+				Play();
+			});
+		}
+
+		public void Play(MusicInfo music)
+		{
+			if (music == null)
+				return;
+
+			currentSong = music;
+			repeat = Game.Settings.Sound.Repeat;
+
+			Sound.PlayMusicThen(music, () =>
+			{
+				if (!installed)
+					return;
+
+				if (!repeat)
+					currentSong = GetNextSong();
+				Play();
+			});
+		}
+
+		public void Play(MusicInfo music, Action onComplete)
+		{
+			if (music == null)
+				return;
+
+			currentSong = music;
+			Sound.PlayMusicThen(music, onComplete);
+		}
+
+		public MusicInfo GetNextSong()
+		{
+			if (!installed)
+				return null;
+
+			var songs = Game.Settings.Sound.Shuffle ? random : music;
+			return songs.SkipWhile(m => m != currentSong)
+				.Skip(1).FirstOrDefault() ?? songs.FirstOrDefault();
+		}
+
+		public MusicInfo GetPrevSong()
+		{
+			if (!installed)
+				return null;
+
+			var songs = Game.Settings.Sound.Shuffle ? random : music;
+			return songs.Reverse().SkipWhile(m => m != currentSong)
+				.Skip(1).FirstOrDefault() ?? songs.Reverse().FirstOrDefault();
+		}
+
+		public void Dispose()
```

Comment:
The `World.Dispose()` method already notifies all relevant traits to stop what they are doing and clean up, so I'd rather we go with that instead of hacking a call to `Playlist` inside `World.Dispose()`. (The current `Sound.StopMusic()` in there was a quick hack to patch things up.)